### DISCUSSION TITLE
websocket usage in non trivial use case

### DISCUSSION THREAD

**Discussion by heuj** [miguelgrinberg/microdot#228] [Q&A]
Hey, while experimenting with microdot and websockets on esp32c3 I came across a question and would be glad to get your feedback.
I checked the basic websocket example: await message and send back, ok so far.

My use case looks a bit different:
On the same websocket I need to 1. wait for incoming messages and 2. periodically send messages to client (w/o client initiating). 

Wondering how this can be realized in elegant way, the normal asyncio wait function allowing to wait for the first coroutine to return (either message received or new message in queue to be sent to client) seems not available on micropython.

Thank you for any hints!

**Comment by miguelgrinberg** [comment]
You could start a background task to do the periodic sends and keep the main request handler blocked on the receive loop. Would that work?

**Comment by heuj** [comment]
This would work, yes! I was unsure where the "ws" object actually lives and how to consider situations in which more than client connects. To avoid this complication I tried to keep it in this one handler like the example

**Reply by miguelgrinberg** [reply]
Each request handler gets its own `ws` object, representing a connection to that specific client. For certain applications it makes sense to maintain a dictionary where the keys are user IDs and the values are the `ws` objects. This would allow you to send messages to any user from any function. There is no rule that says that the `ws` object must only be used in the request handler it was passed as an argument.
